server/src/main/java/com/linecorp/centraldogma/server/internal/api/GitHttpService.java

```diff
@@ -0,0 +1,231 @@
+            return HttpResponse.of(HttpStatus.FORBIDDEN, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Unsupported service: " + service);
+        }
+
+        if (gitProtocol == null || !gitProtocol.contains(VERSION_2_REQUEST)) {
+            return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Unsupported git-protocol: " + gitProtocol);
+        }
+
+        if (!projectApiManager.exists(projectName)) {
+            return HttpResponse.of(HttpStatus.NOT_FOUND, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Project not found: " + projectName);
+        }
+        if (!projectApiManager.getProject(projectName).repos().exists(repoName)) {
+            final String maybeGitSuffixRemoved = maybeRemoveGitSuffix(repoName);
+            if (maybeGitSuffixRemoved == repoName || // We can just check the reference.
+                !projectApiManager.getProject(projectName).repos().exists(maybeGitSuffixRemoved)) {
+                return HttpResponse.of(HttpStatus.NOT_FOUND, MediaType.PLAIN_TEXT_UTF_8,
+                                       "Repository not found: " + repoName);
+            }
+        }
+        return CAPABILITY_ADVERTISEMENT_RESPONSE.toHttpResponse();
+    }
+
+    private static String maybeRemoveGitSuffix(String repoName) {
+        if (repoName.length() > 5 && repoName.endsWith(".git")) {
+            repoName = repoName.substring(0, repoName.length() - 4);
+        }
+        return repoName;
+    }
+
+    // https://www.git-scm.com/docs/gitprotocol-http#_smart_service_git_upload_pack
+    @Post("/{projectName}/{repoName}/git-upload-pack")
+    public HttpResponse gitUploadPack(AggregatedHttpRequest req,
+                                      @Param String projectName, @Param String repoName) {
+        final String gitProtocol = req.headers().get("git-protocol");
+        if (gitProtocol == null || !gitProtocol.contains(VERSION_2_REQUEST)) {
+            return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Unsupported git-protocol: " + gitProtocol);
+        }
+
+        final MediaType contentType = req.headers().contentType();
+        if (contentType == null || !"application/x-git-upload-pack-request".equals(contentType.toString())) {
+            return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Unsupported content-type: " + contentType);
+        }
+
+        if (!projectApiManager.exists(projectName)) {
+            return HttpResponse.of(HttpStatus.NOT_FOUND, MediaType.PLAIN_TEXT_UTF_8,
+                                   "Project not found: " + projectName);
+        }
+        if (!projectApiManager.getProject(projectName).repos().exists(repoName)) {
+            final String maybeGitSuffixRemoved = maybeRemoveGitSuffix(repoName);
```

Comment:
I don't see permission for `xxx.git` is correctly checked with `@RequiresReadPermission`. If the repo name in Central Dogma is `xxx`, `RequiresPermissionDecorator` may fail to find proper permission for the repo.
https://github.com/line/centraldogma/blob/c376db1eaf2ef155bcdf6cd277116e48323fb719/server/src/main/java/com/linecorp/centraldogma/server/internal/api/auth/RequiresPermissionDecorator.java#L88-L89

Reply:
Discussed this with @ikhoon and we've decided remove ".git" suffix if it's specified when creating a repository.

https://github.com/line/centraldogma/issues/956